lib/bundler/definition.rb

```diff
@@ -218,15 +218,42 @@ def index
           dependency_names -= pinned_spec_names(source.specs)
           dependency_names.push(*source.unmet_deps).uniq!
         end
+
+        memoize = proc do |&work|
+          ret = memoize
+          proc do
+            if memoize.equal?(ret)
+              ret = work.call
+            else
+              ret
+            end
+          end
+        end
```

Comment:
I can actually inline it entirely -- I don't like methods because the can be used elsewhere :P